### PR TITLE
Flow tracking

### PR DESCRIPTION
**Pull request by PrasenjitAdhikary**
Proposed model for flows, with a new tree node `flow-metric-tag`:
```
module: open-traffic-generator-flow
  +--rw flows
     +--ro flow* [name]
        +--ro name               -> ../state/name
        +--ro state
        |  +--ro name?             string
        |  +--ro transmit?         boolean
        |  +--ro loss-pct?         otg-types:ieeefloat32
        |  +--ro out-frame-rate?   otg-types:ieeefloat32
        |  +--ro in-frame-rate?    otg-types:ieeefloat32
        |  +--ro out-rate?         otg-types:ieeefloat32
        |  +--ro in-rate?          otg-types:ieeefloat32
        |  +--ro counters
        |     +--ro in-octets?    otg-types:counter64
        |     +--ro in-pkts?      otg-types:counter64
        |     +--ro out-octets?   otg-types:counter64
        |     +--ro out-pkts?     otg-types:counter64
        +--ro flow-metric-tag* [name-value-request]
           +--ro name-value-request    -> ../state/name-value-request
           +--ro state
              +--ro name-value-request?   string
              +--ro metric* [name-value-response]
                 +--ro name-value-response    -> ../state/name-value-response
                 +--ro state
                    +--ro name-value-response?   string
                    +--ro counters
                       +--ro in-octets?    otg-types:counter64
                       +--ro in-pkts?      otg-types:counter64
                       +--ro out-octets?   otg-types:counter64
                       +--ro out-pkts?     otg-types:counter64
```

Assuming the following packet distribution for flow `F1`, some example gNMI query and results are described below:

image

### 1. Only the aggregated per-flow stats:
gNMI query path: “/flows/flow[name=F1]/”
gNMI response:
```
"updates": [
   {
      "Path": "flows/flow[name=f1]",
      "values": {
         "flows/flow": {
            "open-traffic-generator-flow:name": "f1",
            "open-traffic-generator-flow:state": {
               "counters": {                                       // Contains the aggregated per-flow stats
                  "in-octets": "2400",
                  "in-pkts": "300",
                  "out-octets": "2400",
                  "out-pkts": "300"
               },
               "in-frame-rate": "QKAAAA==",
               "name": "f1",
               "out-frame-rate": "QCAAAA==",
               "transmit": true
            }                                                             // Note: disaggregated stats is missing
         }
      }
   }
]
```

### 2. Disaggregated (drill-down) stats for all VLAN IDs:
gNMI query path: “/flows/flow[name=F1]/flow-metric-tag[name-value-request=’vlan’]”
gNMI response:

```
   "updates": [
   {
      "Path": "flows/flow[name=f1]/flow-metric-tag[name-value-request='vlan']",
      "values": {
         "flows/flow/flow-metric-tag": {
            "open-traffic-generator-flow:name-value-request": "vlan",
            "open-traffic-generator-flow:state": {
               "name-value-request": "vlan",
               "metric": [
                  {
                     "name-value-response": "vlan=0x64",        // enumerated to name/value of VLAN ID = 100
                     "counters": {
                        "in-octets": "2000",
                        "in-pkts": "250",
                     }                     
                  },
                  {
                     "name-value-response": "vlan=0xCE",        // enumerated to name/value of VLAN ID = 200
                     "counters": {
                        "in-octets": "400",
                        "in-pkts": "50",
                     }                     
                  }
               ]
         }
      }
   }
]             
```
### 3. Disaggregated stats for VLAN ID = 200:
gNMI query path: “/flows/flow[name=F1]/flow-metric-tag[name-value-request=’vlan=0xCE’]”
gNMI response:
```
   "updates": [
   {
      "Path": "flows/flow[name=f1]/flow-metric-tag[name-value-request='vlan=0xCE']",
      "values": {
         "flows/flow/flow-metric-tag": {
            "open-traffic-generator-flow:name-value-request": "vlan=0xCE",
            "open-traffic-generator-flow:state": {
               "name-value-request": "vlan=0xCE",
               "metric": [
                  {
                     "name-value-response": "vlan=0xCE",        // name/value filtered for VLAN ID = 200
                     "counters": {
                        "in-octets": "400",
                        "in-pkts": "50",
                     }                     
                  }
               ]
         }
      }
   }
]             
```

### 4. Disaggregated stats for all VLAN IDs&Src IP (multi-field tracking):
gNMI query path: “/flows/flow[name=F1]/flow-metric-tag[name-value-request=’vlan&srcip’]”
gNMI response:
```
   "updates": [
   {
      "Path": "flows/flow[name=f1]/flow-metric-tag[name-value-request='vlan&srcip']",
      "values": {
         "flows/flow/flow-metric-tag": {
            "open-traffic-generator-flow:name-value-request": "vlan&srcip",
            "open-traffic-generator-flow:state": {
               "name-value-request": "vlan&srcip",
               "metric": [
                  {
                     "name-value-response": "vlan=0x64&srcip=0x1",   // row for VLAN ID = 100 & Src IP = 10.10.10.1
                     "counters": {
                        "in-octets": "800",
                        "in-pkts": "100",
                     }                     
                  },
                  {
                     "name-value-response": "vlan=0x64&srcip=0x2",   // row for VLAN ID = 100 & Src IP = 10.10.10.2
                     "counters": {
                        "in-octets": "1200",
                        "in-pkts": "150",
                     }                     
                 },
                 {
                     "name-value-response": "vlan=0xCE&srcip=0x1",  // row for VLAN ID = 200 & Src IP = 10.10.10.1
                     "counters": {
                        "in-octets": "160",
                        "in-pkts": "20",
                     }                     
                 },
                 {
                     "name-value-response": "vlan=0xCE&srcip=0x1",  // row for VLAN ID = 200 & Src IP = 10.10.10.2
                     "counters": {
                        "in-octets": "240",
                        "in-pkts": "30",
                     }                     
                 }
               ]
         }
      }
   }
]             
```
**Note:** _Assumption, Metric Tag for Src IP is configured only for the last 8 bits. Hence IP Src address is matched for the last 8 bits appearing in the result._ 

### 5. Disaggregated stats for all Src IPs (VLAN ID is ignored):
gNMI query path: “/flows/flow[name=F1]/flow-metric-tag[name-value-request=’srcip’]”
gNMI response:
```
   "updates": [
   {
      "Path": "flows/flow[name=f1]/flow-metric-tag[name-value-request='srcip']",
      "values": {
         "flows/flow/flow-metric-tag": {
            "open-traffic-generator-flow:name-value-request": "srcip",
            "open-traffic-generator-flow:state": {
               "name-value-request": "srcip",
               "metric": [
                  {
                     "name-value-response": "srcip=0x1",  // row(s) for Src IP = 10.10.10.1
                     "counters": {
                        "in-octets": "960",
                        "in-pkts": "120",             // 100 (for VLAN ID = 100) + 20 (for VLAN ID = 200)
                     }                     
                  },
                  {
                     "name-value-response": "srcip=0x2",  // row(s) for Src IP = 10.10.10.2
                     "counters": {
                        "in-octets": "1440",
                        "in-pkts": "180",              // 150 (for VLAN ID = 100) + 30 (for VLAN ID = 200)
                     }                     
                 }
               ]
         }
      }
   }
]             
```

### CODE REVIEW

File: models/flow/open-traffic-generator-flow.yang [hunk 1]
```diff
@@ -70,6 +70,8 @@ module open-traffic-generator-flow {
             uses flow-counters;
           }
         }
+
+        uses flow-metric-tags;
```
[the remark this answers is not in the document]
Comment (PrasenjitAdhikary) [reply models/flow/open-traffic-generator-flow.yang]: Done.

File: models/flow/open-traffic-generator-flow.yang [hunk 2]
```diff
@@ -158,5 +160,97 @@ module open-traffic-generator-flow {
...
+
+        container state {
+          description
+            "Operational state of the individual flow-metric-tag.";
+
+          leaf name-value-request {
+            type string;
+            description
+              "Encoded string representing one or more metric tag name and corresponding filter value
+              used as request to disaggregate the flow metrics.
```
[the remark this answers is not in the document]
Comment (PrasenjitAdhikary) [reply models/flow/open-traffic-generator-flow.yang]: Renamed all occurrences of disaggregate to enumerate.

File: models/flow/open-traffic-generator-flow.yang [hunk 3]
```diff
@@ -158,5 +160,97 @@ module open-traffic-generator-flow {
...
+              Encoding format:
+                  tag_name=value&tag_name=value
+                  name can be specified multiple times.
+                  value is optional, need to be supplied only for retrieving subset of disaggregated flow metrics
```
[the remark this answers is not in the document]
Comment (PrasenjitAdhikary) [reply models/flow/open-traffic-generator-flow.yang]: Specified explicitly data type for the value.

**Review by hashwini-keysight — APPROVED** [review]

**Comment by greg-dennis** [comment]
Where you say "// Note: disaggregated stats is missing," what do you mean by missing? gNMI requires they would be present for that request, as the tree is currently structured. Do you mean you omitted them to keep the description above simpler or that they would actually be omitted in practice?

**Comment by greg-dennis** [comment]
Well, now I see that in general you want to represent a list that has keys whose data is only produced when requested.

Another option is to produce the full cross-product of all key-value pairs and let users aggregate by whatever dimensions they want themselves. Is producing that full cross product likely to be very expensive?

**Comment by PrasenjitAdhikary** [comment]
> Well, now I see that in general you want to represent a list that has keys whose data is only produced when requested.
> 
> Another option is to produce the full cross-product of all key-value pairs and let users aggregate by whatever dimensions they want themselves. Is producing that full cross product likely to be very expensive?

If multiple fields are tagged for tracking then it will be very costly to produce results for all possible combinations.
For example, the user might have enabled tracking on Src IP, Dst IP, VLAN. If a specific request is not supported, then the result set needs to be prepared for all of the following combinations, always:

1. Src IP Only (all possibilities)
2. Dst IP Only (all possibilities)
3. VLAN Only (all possibilities)
4. Src IP + Dst IP (all possibilities)
5. Src IP + VLAN (all possibilities)
6. Dst IP + VLAN (all possibilities)
7. Src IP + Dst IP + VLAN (all possibility)

Whereas if a user is interested to get results grouped only for the VLAN (or VLAN for some specific value), there is no overhead of producing stats for other fields (which will increase with the increase of tracking fields).
Another issue, Tx stats are available only when the result contains no egress tracking field. In this example, if we consider Src IP & Dst IP are ingress tracking fields and VLAN is an egress tracking field, then the availability of Tx/Rx stats are shown below:

1. Src IP Only: Both Tx & Rx stats
2. Dst IP Only: Both Tx & Rx stats 
3. VLAN Only: Only Rx stats
4. Src IP + Dst IP: Both Tx & Rx stats
5. Src IP + VLAN: Only Rx stats
6. Dst IP + VLAN: Only Rx stats
7. Src IP + Dst IP + VLAN: Only Rx stats

The reason, for an egress field, the implementation does not know how many corresponding packets were generated by the Tx port. For this limitation, drill-down statistics at the lowest level (here  Src IP + Dst IP + VLAN) contains only Rx stats and hence cannot be aggregated at the client side to derive Tx stats for (Src IP + Dst IP) combination.

**Comment by PrasenjitAdhikary** [comment]
> flows/flow[name=f1]"

With this approach, disaggregated stats are produced only if a key value is present for "flow-metric-tag" in the query. If the path is "flows/flow[name=f1]" we are assuming the user is not interested in disaggregated stats and hence only the aggregated stats are produced. Though "disaggregated" stats is structured as a child node under the flow tree, the node will be absent as it contains no value.

**Review by anjan-keysight — APPROVED** [review]
LGTM.

**Comment by greg-dennis** [comment]
@PrasenjitAdhikary we discussed this further, and we would prefer if there were an OTG API for configuring the flow stats disaggregation, the result of which would be to make the disaggregated stats appear. If someone queried the flow container, they would receive both the aggregated and, if configured, the disaggregated stats, too. The reason is that in gNMI, list elements are not supposed to appear only when queried; they are either present and need to be received when querying any container; or they are not present.

**Comment by greg-dennis** [comment]
A couple more comments.

A clarification on https://github.com/open-traffic-generator/models-yang/pull/24#issuecomment-1492265332: I did _not_ mean that all 7 of those combinations would be produced under that option. With that option, only #7 (Src IP + Dst IP + VLAN) would be available through gNMI, and the test/application could use that info to derive all the others as they want using code.

And a clarification on option in https://github.com/open-traffic-generator/models-yang/pull/24#issuecomment-1496280618. If you go this route, you could either (a) allow users to configure multiple disaggregated views at a time (which would require a list); or (b) only allow one view at a time (which would not). In (b), if you wanted multiple views, you would just configure the view, query the stats, then configure the new view and query the stats again -- might be simpler.

**Comment by PrasenjitAdhikary** [comment]
@greg-dennis We are working to address all your comments in the models-yang and in OTG API. Will update you once we have the modified proposal ready.

**Comment by PrasenjitAdhikary** [comment]
> @greg-dennis We are working to address all your comments in the models-yang and in OTG API. Will update you once we have the modified proposal ready.

@greg-dennis Created new PR https://github.com/open-traffic-generator/models-yang/pull/27 to address your comments.

**Pull request merged**